langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/internal/MatchingService.java

```diff
@@ -0,0 +1,111 @@
+
+@Slf4j
+@Builder
+public class MatchingService {
+
+    @NonNull
+    private final String endpoint;
+    @NonNull
+    private final String project;
+    @NonNull
+    private final String location;
+    @NonNull
+    private final String indexEndpointId;
+    @NonNull
+    private final String deployedIndexId;
+    @NonNull
+    private final String indexId;
+    @NonNull
+    private final GcpBlobService gcpBlobService;
+    private final CredentialsProvider credentialsProvider;
+    @Getter
+    @Builder.Default
+    private final boolean returnFullDatapoint = true;
+    @Getter(lazy = true)
+    private final MatchServiceSettings settings = initSettings();
+
+    /**
+     * Add a document to the index
+     *
+     * @param referenceEmbedding the embedding of the document
+     * @param maxResults         the maximum number of results to return
+     * @param minScore           the minimum score to return
+     * @return the list of matching documents
+     */
+    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
+        final IndexDatapoint datapoint = IndexDatapoint
+                .newBuilder()
+                .addAllFeatureVector(referenceEmbedding.vectorAsList())
+                .build();
+        final FindNeighborsRequest.Query query = FindNeighborsRequest.Query
+                .newBuilder()
+                .setDatapoint(datapoint)
+                .setNeighborCount(maxResults)
+                .build();
+        final FindNeighborsRequest request = FindNeighborsRequest
+                .newBuilder()
+                .setIndexEndpoint(IndexEndpointName.of(project, location, indexEndpointId).toString())
+                .setDeployedIndexId(deployedIndexId)
+                .addQueries(query)
+                .setReturnFullDatapoint(returnFullDatapoint)
+                .build();
+
+        try (MatchServiceClient client = MatchServiceClient.create(getSettings())) {
+            final FindNeighborsResponse response = client.findNeighbors(request);
+            return response.getNearestNeighborsList()
+                    .stream()
+                    .flatMap(neighbor -> neighbor.getNeighborsList().stream())
+                    .filter(neighbor -> neighbor.getDistance() >= minScore)
+                    .map(neighbor -> {
+                        final String id = neighbor.getDatapoint().getDatapointId();
+                        final String path = "documents/" + id;
+                        final String content = gcpBlobService.download(path);
+                        final VertexAIDocument document = VertexAIDocument.fromJson(content);
+                        final IndexDatapoint resultDatapoint = neighbor.getDatapoint();
+                        return new EmbeddingMatch<>(neighbor.getDistance(),
+                                id,
+                                Embedding.from(resultDatapoint.getFeatureVectorList()),
```

Comment:
since you made returnFullDatapoint configurable, the check is needed here that the embedding is returned

Reply:
It return an empty array. I think we should also return an empty embedding.

Reply:
That's debatable. Same goes with strings and other empty/null entities. 
Anyway, happy to return null instead of empty if no result.